languages/source.nim

```diff
@@ -100,23 +100,76 @@ const lang* = language:
     FieldAccess(le, IntVal(n))
     At(le, IntVal(n))
 
-  func desugar(a: expr) -> e =
-    # FIXME: the sub-expressions need to be desugared too!
+  func update(a: expr, xfrm: e -> e) -> expr =
```

Comment:
This is really nice, it will help avoid a lot of bugs (missed conditions in otherwise the same traversal) and promote consistency in designing rules language features.